### PR TITLE
Add check for clang-format compliance

### PR DESCRIPTION
**Pull request by marcomagdy**
*Description of changes:*

Add clang-format check to travis.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,95 @@
+---
+Language:        Cpp
```
Comment (JonathanHenson) [inline .clang-format]: why CPP?

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+# PenaltyBreakFirstLessLess: 120
+# PenaltyBreakString: 1000
+# PenaltyExcessCharacter: 1000000
+# PenaltyReturnTypeOnItsOwnLine: 60
+# PointerAlignment: Right
+# ReflowComments:  true
+# SortIncludes:    false
+# SpaceAfterCStyleCast: false
+# SpaceBeforeAssignmentOperators: true
+# SpaceBeforeParens: ControlStatements
```
Comment (JonathanHenson) [inline .clang-format]: don't you want this one?

File: .travis.yml [hunk 3]
```diff
@@ -8,6 +8,14 @@ os:
 #  - osx add osx back when travis fixes their mac os scaling problems.
   - linux
 
+addons:
+  apt:
+    sources:
+    # add PPAs with more up-to-date toolchains
+    - llvm-toolchain-precise-3.8
+    packages:
+    - clang-format-3.8
```
Comment (JonathanHenson) [inline .travis.yml]: if it's possible to move the cppcheck install here, we could lose the sudo and move to sandboxed builds.

**Comment by JonathanHenson** [comment]
Also, I'd like the output of what failed similar to cppcheck

File: include/aws/common/array_list.h [hunk 4]
```diff
@@ -30,107 +30,113 @@ struct aws_array_list {
...
-     */
-    AWS_COMMON_API int aws_array_list_set_at(struct aws_array_list *list, const void *val, size_t index);
-
-    /**
-     * Swap elements at the specified indices.
-     */
-    void aws_array_list_swap(struct aws_array_list *list, size_t a, size_t b);
+/**
+ * Initializes an array list with an array of size initial_item_allocation * item_size. In this mode, the array size
+ * will
```
Comment (JonathanHenson) [inline include/aws/common/array_list.h]: these probably need manual correction.

File: include/aws/common/array_list.h [hunk 5]
```diff
@@ -30,107 +30,113 @@ struct aws_array_list {
...
+/**
+ * Copies the element at the front of the list if it exists. If list is empty, AWS_ERROR_LIST_EMPTY will be raised
+ */
+AWS_COMMON_API int aws_array_list_front(const struct aws_array_list *list, void *val);
+
+/**
+ * Deletes the element at the front of the list if it exists. If list is empty, AWS_ERROR_LIST_EMPTY will be raised.
+ * This call results in shifting all of the elements at the end of the array to the front. Avoid this call unless that
+ * is
+ * intended
```
Comment (JonathanHenson) [inline include/aws/common/array_list.h]: same here.

File: source/encoding.c [hunk 6]
```diff
@@ -25,22 +25,20 @@ static const char base64_encoding_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghi
...
-            41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
-            0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD };
+    64,   0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD, 0xDD,
```
Comment (JonathanHenson) [inline source/encoding.c]: this one is wrong, they're laid out that way for easy navigation (each line is 16 bytes)

File: source/encoding.c [hunk 7]
```diff
@@ -196,8 +194,7 @@ int aws_base64_compute_decoded_len(const char *input, size_t len, size_t *decode
 
     if (len >= 2 && input[len - 1] == '=' && input[len - 2] == '=') { /*last two chars are = */
         padding = 2;
-    }
-    else if (input[len - 1] == '=') { /*last char is = */
+    } else if (input[len - 1] == '=') { /*last char is = */
```
Comment (JonathanHenson) [inline source/encoding.c]: we need an exception on this rule.

File: .clang-format (hunk 1 above)
Comment (marcomagdy) [reply .clang-format]: clang-format won't work with just C

File: .clang-format (hunk 2 above)
Comment (marcomagdy) [reply .clang-format]: This is for space before parens even when calling a function:
```c
int x = foo ();
           ^
```

File: .travis.yml (hunk 3 above)
Comment (marcomagdy) [reply .travis.yml]: I'll try that in a separate PR

File: include/aws/common/array_list.h (hunk 4 above)
Comment (marcomagdy) [reply include/aws/common/array_list.h]: clang-format doesn't realize there's a brace that's #ifdef-ed. 
I think it's ok to let this one slip.

File: source/encoding.c (hunk 6 above)
Comment (marcomagdy) [reply source/encoding.c]: Fixed

File: source/encoding.c (hunk 7 above)
Comment (marcomagdy) [reply source/encoding.c]: done

**Pull request merged**